DataStats.ecl

```diff
@@ -1,49 +1,53 @@
-IMPORT $ AS LogisticRegression;
-IMPORT LogisticRegression.Types AS Types;
-IMPORT LogisticRegression.Constants AS Constants;
+﻿IMPORT $ AS GLM;
+IMPORT GLM.Types AS Types;
+IMPORT GLM.Constants AS Constants;
+IMPORT GLM.Family;
 IMPORT ML_Core.Types AS Core_Types;
 // convenient aliases
 Data_Info := Types.Data_Info;
 Field_Desc := Types.Field_Desc;
 NumericField := Core_Types.NumericField;
-DiscreteField:= Core_Types.DiscreteField;
 // working definitions
 Flat_Field_Desc := RECORD(Types.Field_Desc)
   Core_Types.t_work_item wi;
 END;
 
 /**
  * Information about the datasets.  Without details the range
- *for the x and y (independent and dependent) columns.  Note that
- *a column of all zero values cannot be distinguished from a missing
- *column.
+ * for the x and y (independent and dependent) columns.  Note that
+ * a column of all zero values cannot be distinguished from a missing
+ * column.
  * When details are requested, the cardinality, minimum, and maximum
- *values are returned.  A zero cardinality is returned when the field
- *cardinality exceeds the Constants.limit_card value.
+ * values are returned.  A zero cardinality is returned when the field
+ * cardinality exceeds the Constants.limit_card value.
  * @param indep data set of independent variables
  * @param dep data set of dependent variables
  * @param field_details Boolean directive to provide field level info
+ * @param fam A module defining the error distribution and link of the response.
  * @returns a data set of information on each work item
  */
 EXPORT DATASET(Types.Data_Info)
        DataStats(DATASET(Core_Types.NumericField) indep,
-                  DATASET(Core_Types.DiscreteField) dep,
-                  BOOLEAN field_details=FALSE) := FUNCTION
+                  DATASET(Core_Types.NumericField) dep,
+                  BOOLEAN field_details=FALSE,
+                  Family.FamilyInterface fam=Family.Gaussian) := FUNCTION
   // assemble details for independent and dependent data
   // dependent details, treat as roughly grouped by work item
   l1_dep := GROUP(dep(field_details), wi, LOCAL);
   l1_dep_srt := SORT(l1_dep, number , value);
-  l1_dep_grp := GROUP(l1_dep_srt, number);
+  l1_dep_grp := GROUP(l1_dep_srt, number, LOCAL);
```

Comment:
@johnholt could you tell me where I'm going wrong, I'd like to see if there's a way to get a similar speedup in a similar way if this approach doesn't work. My thinking was that:

```l1_dep := GROUP(dep(field_details), wi, LOCAL);``` will put all records with the same wi on the same node, then ```l1_dep_srt := SORT(l1_dep, number , value);``` will keep that grouping and distribution (as it's implicitly local) but will sort within each grouping by value and number. Finally, ```l1_dep_grp := GROUP(l1_dep_srt, number, LOCAL)``` would leave the dataset grouped by wi and number (and sorted by value within each group). So that there may be records with the same number on different nodes, but only if they are from different work items, which is okay.

Let me know where the mistake is and try find another solution.

Reply:
I think I missed the earlier I1_dep definition that groups the data by wi locally.  My comment is incorrect.

However, the original sequence is not what you describe.  I1_dep does not move any records across nodes or even within the nodes.  So, I1_dep can have more than one node with records from the same work_item, AND, there can be more than 1 group for each work_item value on the same node.

Since I1_dep is already grouped, the LOCAL flag on I1_dep_grp does nothing.  I think that it should have generated a warning though.

The idea in the original sequence was to perform a local reduction to unique strings, and then if there were not too many instances determine a global list of unique strings.